src/middleware/normalize.rs

```diff
@@ -0,0 +1,93 @@
+//! `Middleware` to normalize request's URI
+
+use regex::Regex;
+use actix_service::{Service, Transform};
+use futures::future::{self, FutureResult};
+
+use crate::service::{ServiceRequest, ServiceResponse};
+
+lazy_static::lazy_static! {
+    static ref MERGE_SLASH: Regex = Regex::new("//+").unwrap();
+}
+
+#[derive(Default, Clone, Copy)]
+/// `Middleware` to normalize request's URI in place
+///
+/// Performs following:
+///
+/// - Merges double slashes into one.
+/// - Removes trailing slash.
+pub struct InPlace;
+
+impl<S> Transform<S> for InPlace
+where
+    S: Service<Request = ServiceRequest, Response = ServiceResponse>,
+{
+    type Request = ServiceRequest;
+    type Response = ServiceResponse;
+    type Error = S::Error;
+    type InitError = ();
+    type Transform = InPlaceNormalization<S>;
+    type Future = FutureResult<Self::Transform, Self::InitError>;
+
+    fn new_transform(&self, service: S) -> Self::Future {
+        future::ok(InPlaceNormalization { service })
+    }
+}
+
+pub struct InPlaceNormalization<S> {
+    service: S,
+}
+
+impl<S> Service for InPlaceNormalization<S>
+where
+    S: Service<Request = ServiceRequest, Response = ServiceResponse>,
+{
+    type Request = ServiceRequest;
+    type Response = ServiceResponse;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self) -> futures::Poll<(), Self::Error> {
+        self.service.poll_ready()
+    }
+
+    fn call(&mut self, mut req: ServiceRequest) -> Self::Future {
+        let head = req.head_mut();
+
+        let path = head.uri.path();
+        let original_len = path.len();
+        let path = MERGE_SLASH.replace_all(path, "/");
+        let path = path.trim_end_matches('/');
```

Comment:
i am not sure about this. what case does this cover?

Reply:
To trim trailing slash

Reply:
So, that means I wouldn’t be able to have resources with trailing slash?

Reply:
Yes, would you like to make it optional?

General idea is that it would equally normalize all paths so that they would have the same style.
After all, is there a difference between `/path` and `/path/`?

Reply:
That is no option. This should not happen. Remove it. And add more tests

Reply:
Ok